controller/service.go

```diff
@@ -155,6 +156,35 @@ func (c *Controller) getPreviewAndActiveServices(r *v1alpha1.Rollout) (*corev1.S
 	return previewSvc, activeSvc, nil
 }
 
+func (c *Controller) reconcileCanaryService(r *v1alpha1.Rollout) error {
+	if r.Spec.Strategy.CanaryStrategy != nil && r.Spec.Strategy.CanaryStrategy.CanaryService != "" {
+		svc, err := c.servicesLister.Services(r.Namespace).Get(r.Spec.Strategy.CanaryStrategy.CanaryService)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				msg := fmt.Sprintf(conditions.ServiceNotFoundMessage, r.Spec.Strategy.CanaryStrategy.CanaryService)
+				c.recorder.Event(r, corev1.EventTypeWarning, conditions.ServiceNotFoundReason, msg)
+				newStatus := r.Status.DeepCopy()
+				cond := conditions.NewRolloutCondition(v1alpha1.RolloutProgressing, corev1.ConditionFalse, conditions.ServiceNotFoundReason, msg)
+				conditions.SetRolloutCondition(newStatus, *cond)
+				c.persistRolloutStatus(r, newStatus, &r.Spec.Paused)
```

Comment:
I'm wondering if we should abstract this logic into a separate method because we have this code repeated 3 times. Thoughts?

Reply:
good catch! done